XMonad/Prompt/DotDesktop.hs

```diff
@@ -0,0 +1,146 @@
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE LambdaCase #-}
+module XMonad.Prompt.DotDesktop
+    ( appLaunchPrompt
+    ) where
+
+import XMonad ( spawn, io, X )
+import XMonad.Prompt ( mkXPrompt, XPConfig(searchPredicate) )
+import XMonad.Prompt.Shell ( Shell(Shell), split )
+import XMonad.Prompt.DotDesktopParser ( runDotDesktopParser )
+
+import qualified Data.Map as M
+import Control.Applicative ( Alternative((<|>)) )
+import Control.Monad (filterM)
+import Control.Monad.Except
+    ( runExceptT, ExceptT (ExceptT), liftEither )
+import Control.Exception ( try, Exception )
+import Data.Functor ( (<&>) )
+import Data.List ( isSuffixOf, dropWhileEnd )
+import Data.Maybe ( fromMaybe, maybeToList, listToMaybe )
+import System.Directory (listDirectory, doesDirectoryExist)
+import System.Environment ( lookupEnv )
+import System.FilePath ((</>))
+
+import Data.Char (isSpace)
+import Data.Either (rights, lefts)
+import XMonad.Prelude (join)
+
+isDotDesktop :: FilePath -> Bool
+isDotDesktop = isSuffixOf ".desktop"
+
+trimWhitespace :: String -> String
+trimWhitespace = dropWhileEnd isSpace . dropWhile isSpace
+
+cmdFilter :: String -> String  -- fixme future do something other than dropping these
+cmdFilter ('%':'f':xs) = cmdFilter xs
+cmdFilter ('%':'F':xs) = cmdFilter xs
+cmdFilter ('%':'u':xs) = cmdFilter xs
+cmdFilter ('%':'U':xs) = cmdFilter xs
+cmdFilter ('%':'c':xs) = cmdFilter xs
+cmdFilter ('%':'k':xs) = cmdFilter xs
+cmdFilter ('%':'i':xs) = cmdFilter xs
+cmdFilter ('%':'%':xs) = '%' : cmdFilter xs
+cmdFilter (x:xs) = x : cmdFilter xs
+cmdFilter "" = ""
```

Comment:
Well this doesn't look fun :)  I'm not familiar with `.desktop` files at all; what exactly are you doing here?

Reply:
I want to just execute the .desktop command specified, without arguments.
Per the spec here:
https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
there are many field codes for arguments which I cannot pass directly to the shell to execute, so I just remove them.

Reply:
You might want to see if using `xdg-open` to run it gives you more options.

Reply:
I do not understand how I would use xdg-open?  Could you provide an example?

Just for context, here is an example from `gimp.desktop`:
```
[Desktop Entry]
Version=1.0
Type=Application
Name=GNU Image Manipulation Program
...
Exec=gimp-2.10 %U
...
```
Basically the user could type `image`, fuzzyMatch finds and displays `GNU Image Manipulation Program`, and if it is selected, the Exec is `gimp-2.10 %U`.  This code filters out the %U and passes `gimp-2.10` to the shell to launch.

Reply:
`gtk-launch` or `kioclient5 exec` can be used to run a .desktop file: https://askubuntu.com/a/1114798/950919

Reply:
ah, I get it.  I did not realize you could just xdg-open the .desktop file.

What do all think?  Just use the exec with the cmd filter or use xdg-open or gtk-launch?

Reply:
Well it's a bit silly though. `xdg-open` doesn't run the desktop file, it opens it in some app. `gtk-launch` works, but it expects just `gimp.desktop`, not `/usr/share/applications/gimp.desktop`. `kioclient3 exec`, on the other hand, only works with `/usr/share/applications/gimp.desktop` but not with `gimp.desktop`. It's just chaos. :-(

Reply:
`gtk-launch` is (well, at least on my distro) installed with the `gtk` package itself, so chances are users will definitely have this installed.

Will the fact that it goes into `/usr/share/applications` by itself be a problem?  Out of the current paths that the code searches (`~/.local/share/applications`, `/usr/share/applications`, and `/usr/local/share/applications`) on the first one seems not to get searched by `gtk-launch`.

Reply:
`gtk-launch` does look into `~/.local/share/applications` here.

Reply:
Mh, doesn't work on my end; a `.desktop` file that's only in `~/.local/share/applications` but not in any of the other directories will fail to be found

Reply:
That's really weird, I looked into the source of gtk/glib and I don't see why it wouldn't look there. Maybe try stracing it?

Reply:
Oh indeed, `strace` tells me that the directory *is* searched.  Upon closer inspection, the `.desktop` file contains a non-existant executable; instead of telling me that, `gtk-launch` told me that the `.desktop` file couldn't be found, however ._.